returnn/search.py

```diff
@@ -14,6 +14,178 @@
+        :param tk.Path|str|None returnn_python_exe: path to the RETURNN executable (python binary or launch script)
+        :param tk.Path|str|None returnn_root: path to the RETURNN src folder
+        """
+        self.search_data = search_data
+
+        self.model_checkpoint = model_checkpoints[
+            epoch.get() if isinstance(epoch, tk.Variable) else epoch
+        ]
+
+        self.returnn_python_exe = (
+            returnn_python_exe
+            if returnn_python_exe is not None
+            else gs.RETURNN_PYTHON_EXE
+        )
+
+        self.returnn_root = (
+            returnn_root if returnn_root is not None else gs.RETURNN_ROOT
+        )
+
+        self.out_search_file = self.output_path("search_out")
+
+        self.returnn_config = self.create_returnn_config(
+            returnn_config, log_verbosity, device
+        )
+
+        self.out_returnn_config_file = self.output_path("returnn.config")
+
+        self.rqmt = {
+            "gpu": 1 if device == "gpu" else 0,
+            "cpu": cpu_rqmt,
+            "mem": mem_rqmt,
+            "time": time_rqmt,
+        }
+
+    def tasks(self):
+        yield Task("create_files", mini_task=True)
+        yield Task("run", resume="run", rqmt=self.rqmt)
+
+    def create_files(self):
+        config = self.returnn_config
+
+        config.config["load"] = self.model_checkpoint.ckpt_path
+        config.write(self.out_returnn_config_file.get_path())
+
+        with open("rnn.sh", "wt") as f:
+            f.write(
+                "#!/usr/bin/env bash\n%s"
+                % " ".join(
+                    [
+                        tk.uncached_path(self.returnn_python_exe),
+                        os.path.join(tk.uncached_path(self.returnn_root), "rnn.py"),
+                        self.out_returnn_config_file.get_path(),
+                    ]
+                )
+            )
+
+        os.chmod(
+            "rnn.sh",
+            stat.S_IRUSR
+            | stat.S_IRGRP
+            | stat.S_IROTH
+            | stat.S_IWUSR
+            | stat.S_IXUSR
+            | stat.S_IXGRP
+            | stat.S_IXOTH,
+        )
+
+        # check here if model actually exists
+        assert os.path.exists(
+            tk.uncached_path(self.model_checkpoint.index_path)
+        ), "Provided model does not exists: %s" % str(self.model_checkpoint)
+
+    def run(self):
+        call = [
+            tk.uncached_path(self.returnn_python_exe),
+            os.path.join(tk.uncached_path(self.returnn_root), "rnn.py"),
+            self.out_returnn_config_file.get_path(),
+        ]
+        sp.check_call(call)
+
+    def create_returnn_config(self, returnn_config, log_verbosity, device):
+        """
+        Creates search RETURNN config
+        :param ReturnnConfig returnn_config: object representing RETURNN config
+        :param int log_verbosity: RETURNN log verbosity
+        :param str device: RETURNN device, cpu or gpu
+        :return: search ReturnnConfig
```

Comment:
```suggestion
        :rtype: ReturnnConfig
```
Btw, should we keep using the docstring to specify types?
Or rather use Python type annotations? (I don't find whether we discussed or decided on this.)

Reply:
yeah also not sure about this. i think it is not yet discussed.